btc_light_client.go

```diff
@@ -48,57 +48,73 @@ func (lc *BTCLightClient) FindPreviousCheckpoint() (blockchain.HeaderCtx, error)
 	return nil, nil
 }
 
-func (lc *BTCLightClient) SetHeader(height int64, header wire.BlockHeader) error {
-	return lc.btcStore.SetHeader(height, header)
-}
-
 // We assume we always insert valid header. Acctually, Cosmos can revert a state
 // when module return error so this assumtion is reasonable
-func (lc *BTCLightClient) InsertHeaders(headers []wire.BlockHeader) error {
-	latestHeight := lc.btcStore.LatestHeight()
+func (lc *BTCLightClient) InsertHeaders(header wire.BlockHeader) error {
 
-	return lc.insertHeaderStartAtHeight(uint64(latestHeight), headers)
-}
+	if lb := lc.btcStore.LightBlockByHash(header.BlockHash()); lb != nil {
+		return errors.New("Parent block not found")
+	}
+	
+	previousBlockHash := header.PrevBlock
+	previousBlock := lc.btcStore.LightBlockByHash(previousBlockHash)
+	if previousBlock == nil {
+		return errors.New("Block doesn't belong to any fork!")
+	}
+
+	// we need to handle 2 cases:
+	// extend the exist fork
+	if lc.btcStore.RemindFork(previousBlockHash) {
+		previousHeader := previousBlock.Header
+		if err := lc.CheckHeader(previousHeader, header); err != nil {
+			return err
 
-func (lc *BTCLightClient) insertHeaderStartAtHeight(startHeight uint64, headers []wire.BlockHeader) error {
-	height := startHeight + 1
-	for i, header := range headers {
-		if err := lc.CheckHeader(header); err != nil {
-			return NewInvalidHeaderErr(header.BlockHash().String(), i)
 		}
 
-		// override a height
-		lc.SetHeader(int64(height), header)
-		height++
+		lc.btcStore.AddBlock(previousBlock, header)
+		lc.btcStore.SetLatestBlockOnFork(previousBlockHash, false)
+		lc.btcStore.SetLatestBlockOnFork(header.BlockHash(), true)
+		return nil
 	}
-	return nil
+
+	// create a new fork
+	parent := lc.btcStore.LightBlockByHash(previousBlockHash)
+	return lc.CreateNewFork(parent, header)
 }
 
-func (lc *BTCLightClient) sumTotalWork(startBlock *LightBlock, headers []wire.BlockHeader) *big.Int {
-	totalWork := startBlock.TotalWork
-	for _, header := range headers {
-		totalWork.Add(totalWork, blockchain.CalcWork(header.Bits))
+func (lc *BTCLightClient) extractFork(bh chainhash.Hash) ([]*LightBlock, error) {
+	checkpoint := lc.btcStore.LatestCheckPoint()
```

Comment:
There can be few heads. Example, where we have 3 active forks, with 3 latest blocks (b3, c3, d3):
```
b1 <- b2  <- b3
   |- c2  <- c3
   |- d2'  <- d3
```